lib/shared/bucketing-assembly-script/assembly/eventQueue/requestPayloadManager.ts

```diff
@@ -101,32 +101,59 @@ export class RequestPayloadManager {
             }
         }
 
-        // Generate defaulted agreagate user as our events APIs require a user / user_id
+        // Generate defaulted aggregate user as our events APIs require a user / user_id
         const dvcUser = new DVCPopulatedUser(new DVCUser(
             user_id, null, null, null, null, NaN, null, null, null, null
         ))
         this.addEventsToPendingPayloads(new UserEventsBatchRecord(dvcUser, aggEvents))
     }
 
+    /**
+     * Find existing FlushPayload that contains records for the user_id,
+     * or return the smallest FlushPayload or a new FlushPayload
+     */
+    private getFlushPayloadForUser_id(user_id: string): FlushPayload {
+        const keys = this.pendingPayloads.keys()
+        let smallestPayload: FlushPayload | null = null
+
+        for (let i=0; i < keys.length; i++) {
+            const key = keys[i]
+            const payload = this.pendingPayloads.get(key)
+            const recordForUser = payload.recordForUser(user_id)
+            const payloadEventCount = payload.eventCount()
+            if (recordForUser && payloadEventCount < this.chunkSize) {
+                this.pendingPayloads.delete(key)
+                return payload
+            }
+            if (payloadEventCount < this.chunkSize
+                && (!smallestPayload || smallestPayload.eventCount() < payloadEventCount)) {
+                smallestPayload = payload
+            }
+        }
+
+        if (smallestPayload) {
+            this.pendingPayloads.delete(smallestPayload.payloadId)
+            return smallestPayload
+        } else {
+            return new FlushPayload([])
+        }
+    }
+
     /**
      * Chunk up UserEventsBatchRecord's into payloads of size: this.chunkSize
      */
     private addEventsToPendingPayloads(record: UserEventsBatchRecord): void {
-        let flushPayload = new FlushPayload([])
+        let flushPayload = this.getFlushPayloadForUser_id(record.user.user_id)
 
-        while (record.events.length > this.chunkSize) {
-            const batchRecord = new UserEventsBatchRecord(
-                record.user,
-                record.events.splice(0, this.chunkSize)
-            )
-            flushPayload.records.push(batchRecord)
+        while (record.events.length > (this.chunkSize - flushPayload.eventCount())) {
```

Comment:
it would be clearer to me if this just looped until record.events was empty (or some similar condition asuming you stop mutating that array) and then the contents of the loop would have an `if-else` check for whether to use an existing payload or create a new one

Reply:
yea that makes sense.